### PR TITLE
api-client

### PR DESCRIPTION
**Pull request by koonpeng**
## What's new

A simple api client library that can be used on node or the browser. Also did some organizing and linting improvements, particularly, there is now a base eslint config is shared between the js/ts packages, with each package extending it with it's own config as needed.

other changes:

* mixins and models are now generated at build time, committng them was a mistake as it may results in the dev's version of rmf being tied to the models. It is problematic because this version may be from wip feature branches or out of tree checkout of rmf.

## Self-checks

- [ ] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [ ] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

### CODE REVIEW

File: packages/api-server/ros_tortoise/__main__.py [hunk 1]
```diff
@@ -138,18 +139,24 @@ def main():
...
         mixins = gen_mixin(pkg_spec)
+        sha1.update(mixins.encode())
         outfile = f"{args.output}/{pkg}_mixins.py"
         with open(outfile, "w") as f:
             f.write(mixins)
         print(outfile)
 
+    rev = base64.b32encode(sha1.digest())
+    with open(f"{args.output}/rev", "bw") as f:
+        f.write(rev)
```
Comment (matiasbavera) [inline packages/api-server/ros_tortoise/__main__.py:158]: On models migration to the database, on prod, should we save the rev's hash in case we want to replicate models migration on the localhost, right?  

Is that the idea?

**Review by matiasbavera — APPROVED** [review]
LGTM, left a couple of comments. I tested and it's working correctly.

File: packages/api-server/ros_tortoise/__main__.py (hunk 1 above)
Comment (koonpeng) [reply packages/api-server/ros_tortoise/__main__.py:158]: The initial plan was to provide an easy way for build system to do change detection, but api client is not using a build system now so it is not used.

The rev only tracks and the mixins and not the models so we cant use it for any prod work.

File: packages/api-client/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,71 @@
+{
+  "compilerOptions": {
+    /* Visit https://aka.ms/tsconfig.json to read more about this file */
+
+    /* Basic Options */
+    // "incremental": true,                         /* Enable incremental compilation */
+    "target": "es5",                                /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
+    "module": "commonjs",                           /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/api-client/tsconfig.json:8]: I actually just used the default, ideally it should work for both browser and node. Im not really sure whats the best way to do this, should we build for both env, or distribute the source ts for webpack to do its own bundling?

[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/api-client/tsconfig.json:8]: so the correct way to do this is to have a `main`, `module`, `browser` and other fields in package.json, webpack looks at those fields to determine how to bundle. We should be able to keep using `commonjs` and let webpack convert it for browser, we might want to add a devDependencies on webpack and publish the minified bundle if we ever decide to publish this to npmjs
